lib/Conversion/TorchToLinalg/Linear.cpp

```diff
@@ -228,49 +232,151 @@ class ConvertAtenMatmulOp : public OpConversionPattern<AtenMatmulOp> {
+            castIndexToInt64(rewriter, loc, lhsBroadcastToShape[i]);
+        rhsBroadcastToShape[i] =
+            castIndexToInt64(rewriter, loc, rhsBroadcastToShape[i]);
+      }
+
+      // Broadcast the batch dimensions of both the matrices.
+      Value broadcastedLhs, broadcastedRhs;
+      if (failed(torch_to_linalg::broadcastToGivenShape(
+              op, rewriter, lhs, lhsBroadcastToShape, broadcastedLhs))) {
+        return rewriter.notifyMatchFailure(
+            op, "unable to perform broadcast operation");
+      }
+      if (failed(torch_to_linalg::broadcastToGivenShape(
+              op, rewriter, rhs, rhsBroadcastToShape, broadcastedRhs))) {
+        return rewriter.notifyMatchFailure(
+            op, "unable to perform broadcast operation");
+      }
+
+      // Check if the result of the matrix multiplication has more than one
+      // dynamic batch dimensions.
+      ArrayRef<int64_t> resultShapeInt = resultType.getShape();
+      int64_t count = 0;
+      bool multipleDynamicBatchDims = false;
+      for (int64_t dim : resultShapeInt) {
+        if (dim == -1 && ++count > 1) {
+          multipleDynamicBatchDims = true;
+          break;
+        }
+      }
+
+      // TODO: Lowering to `linalg.BatchMatmul` is only possible when there is
+      // at most one dynamic batch dimension due to limited support of the
+      // `tensor.ExpandShape` op.
+      if (!multipleDynamicBatchDims) {
+        // Collapse the batch dimensions into one dimension. The resultant rank
+        // will always be 3.
+        SmallVector<ReassociationIndices> reassociation(3);
+        for (unsigned i = 0, j = 0; i < maxRank; i++) {
+          if (i >= batchRank)
+            j++;
+          reassociation[j].push_back(i);
+        }
+        Value collapsedLhs = rewriter.create<tensor::CollapseShapeOp>(
+            op->getLoc(), broadcastedLhs, reassociation);
+        Value collapsedRhs = rewriter.create<tensor::CollapseShapeOp>(
+            op->getLoc(), broadcastedRhs, reassociation);
+
+        // Compute the result shape after collapsing the batch dimensions.
+        SmallVector<Value> collapsedResultShape;
+        collapsedResultShape.push_back(broadcastedBatchShape[0]);
+        for (unsigned i = 1; i < batchRank; i++) {
+          collapsedResultShape[0] = rewriter.createOrFold<arith::MulIOp>(
+              loc, collapsedResultShape[0], broadcastedBatchShape[i]);
+        }
+        collapsedResultShape.push_back(lhsDim0);
+        collapsedResultShape.push_back(rhsDim1);
+        SmallVector<OpFoldResult> updatedCollapseResultShape =
+            getAsOpFoldResult(collapsedResultShape);
+
+        Value initTensor = rewriter.create<linalg::InitTensorOp>(
+            loc, updatedCollapseResultShape, elementType);
+        Value c0 = rewriter.create<arith::ConstantOp>(
+            loc, rewriter.getZeroAttr(elementType));
+        Value zeroTensor =
+            rewriter.create<linalg::FillOp>(loc, c0, initTensor).getResult(0);
```

Comment:
nit: you can use `createZeroInitTensor`

Reply:
I did not use this because here the sizes is a SmallVector<OpFoldResult> which does not fit there.